### PR TITLE
Cmake: build in build dir

### PR DESCRIPTION
**Pull request by daleglass**
This makes it so that resources and jsdoc are generated directly into the build directory.

After this PR, the only generated file in the source dir that remains is `tools/jsdoc/package-lock.json`, which might be correct (it says it should go in the source tree) and should be harmless.

Testing:

* It should build correctly, of course
* Interface should start correctly
* resources.rcc and resources.qrc should be generated in the build directory
* `git clean -n -x` in the source dir should only return `tools/jsdoc/package-lock.json` -- this was left alone because I understand that file is supposed to be in the repository.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Alpha-PR943-f761e20.tar_d848d082.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Alpha-PR943-f761e20.tar_d848d082.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Alpha-PR943-f761e20_d848d082.sh

File: hifi_qt.py [hunk 1]
```diff
@@ -73,22 +133,36 @@ def __init__(self, args):
...
             print("Machine     : " + platform.machine())
             raise Exception('UNKNOWN OPERATING SYSTEM!!!')
 
+    def showQtBuildInfo(self):
+        print("")
+        print("It's also possible to build Qt for your distribution, please see the documentation at:")
+        print("https://github.com/vircadia/vircadia/tree/master/tools/qt-builder")
+        print("")
+        print("Alternatively, you can try building against the system Qt by setting the VIRCADIA_USE_SYSTEM_QT environment variable.")
+        print("You'll need to install the development packages, and to have Qt 5.15.0 or newer. ")
```
Comment (ctrlaltdavid) [inline hifi_qt.py:164]:
```suggestion
        print("You'll need to install the development packages, and to have Qt 5.15.0 or later. ")
```
Perhaps.

File: tools/jsdoc/plugins/hifiJSONExport.js [hunk 2]
```diff
@@ -1,8 +1,22 @@
 exports.handlers = {
     processingComplete: function(e) {
         const pathTools = require('path');
-        var outputFolder = pathTools.join(__dirname, '../out');
+        var outputFolder;
         var doclets = e.doclets.map(doclet => Object.assign({}, doclet));
+
+        var argv = process.argv;
+        for(var i=0;i<argv.length;i++) {
+            if ( argv[i] == '-d' && i <= argv.length+1) {
```
Comment (ctrlaltdavid) [inline tools/jsdoc/plugins/hifiJSONExport.js:9]:
```suggestion
        for (var i=0; i < argv.length; i++) {
            if (argv[i] === '-d') {
```
` i <= argv.length + 1` will always be `true`.

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Alpha-PR943-2f0c304.tar_e66f24d6.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Alpha-PR943-2f0c304.tar_e66f24d6.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Alpha-PR943-2f0c304_e66f24d6.sh

**Comment by HifiExperiments** [comment]
does this address https://github.com/vircadia/vircadia/issues/670

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff.tar_f2932e65.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff.tar_f2932e65.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff_f2932e65.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff_f2932e65.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Quest-Alpha-PR943-f15f1ff_f2932e65.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff_f2932e65.apk

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff.tar_f2932e65.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff.tar_f2932e65.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff_f2932e65.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff_f2932e65.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Quest-Alpha-PR943-f15f1ff_f2932e65.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-f15f1ff_f2932e65.apk

**Comment by ctrlaltdavid** [comment]
OSX build is failing with the following:
```
/usr/local/Cellar/cmake/3.19.4/bin/cmake -E copy_directory /Users/runner/work/vircadia/build/interface/tools/jsdoc /Users/runner/work/vircadia/build/interface/Release/interface.app/Contents/MacOS/../Resources/jsdoc
Error copying directory from "/Users/runner/work/vircadia/build/interface/tools/jsdoc" to "/Users/runner/work/vircadia/build/interface/Release/interface.app/Contents/MacOS/../Resources/jsdoc".
make: *** [interface_buildpart_0] Error 1
Command /bin/sh failed with exit code 2
** BUILD FAILED **
```

**Comment by daleglass** [comment]
Okay, this should be functional now!

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-748b33e-.tar_7cb8c535.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-748b33e-.tar_7cb8c535.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-748b33e-_7cb8c535.sh

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-748b33e_7cb8c535.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Quest-Alpha-PR943-748b33e_7cb8c535.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-748b33e-_7cb8c535.exe

File: tools/jsdoc/plugins/hifiJSONExport.js [hunk 3]
```diff
@@ -1,8 +1,22 @@
 exports.handlers = {
     processingComplete: function(e) {
         const pathTools = require('path');
-        var outputFolder = pathTools.join(__dirname, '../out');
+        var outputFolder;
         var doclets = e.doclets.map(doclet => Object.assign({}, doclet));
+
+        var argv = process.argv;
+        for(var i=0; i<argv.length; i++) {
```
Comment (ctrlaltdavid) [inline tools/jsdoc/plugins/hifiJSONExport.js:8]:
```suggestion
        for (var i = 0; i < argv.length; i++) {
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-7ffe519-.tar_41924980.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-7ffe519-_41924980.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-7ffe519-.tar_41924980.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-7ffe519-_41924980.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-2733165-_c3f649a5.dmg

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Interface-PR943-2733165-_c3f649a5.dmg

**Comment by sethalves** [comment]
```
cd /home/seth/src/vircadia/build/interface && /usr/local/bin/cmake -E copy_if_different /home/seth/src/vircadia/interface/resources/serverless/redirect.json /home/seth/src/vircadia/build/interface/resources/serverless/redirect.json
cd /home/seth/src/vircadia/build/interface && /usr/local/bin/cmake -E copy_directory  /home/seth/src/vircadia/build/interface/jsdoc
CMake Error: cmake version 3.21.0-rc2
Usage: /usr/local/bin/cmake -E <command> [arguments...]
```

**Comment by daleglass** [comment]
@sethalves That's a new one, could you post a full log somewhere, including any cmake arguments?

**Comment by sethalves** [comment]
I will post a log, but I think it's probably the same thing that the autobuilder is hitting:

https://github.com/vircadia/vircadia/pull/943/checks?check_run_id=2979649796

**Comment by sethalves** [comment]
The trouble is that I didn't have npm installed, and in tools/CMakeLists.txt :

if (NPM_EXECUTABLE)
  add_subdirectory(jsdoc)
  set_target_properties(jsdoc PROPERTIES FOLDER "Tools")
endif()

so the CMakeLists.txt in jsdoc didn't get processed, and `JSDOC_OUTPUT_PATH` ended up not being set.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Interface-PR943-99cf0c3-_4152e1ba.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-99cf0c3-_4152e1ba.dmg

**Comment by daleglass** [comment]
@sethalves Please retest!

Build is currently failing in the upload script, so ignore that. This version should work both with and without NPM.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Interface-PR943-5358b7b-_276a0506.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-5358b7b-.tar_276a0506.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-5358b7b-_276a0506.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-5358b7b-.tar_276a0506.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-5358b7b-_276a0506.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Quest-Alpha-PR943-5358b7b_276a0506.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-5358b7b_276a0506.apk

**Comment by ctrlaltdavid** [comment]
Windows builds are failing for me, too: The build locks up in Visual Studio with CPU process still being consumed. I cancel the build and restart the build and it gets a bit further. Repeat a few times and eventually it gets to the end but fails with errors: `error C2065: 'M_PI': undeclared identifier`.

[M_PI errors.txt](https://github.com/vircadia/vircadia/files/6812312/M_PI.errors.txt)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-8ea3263-_32c3725b.dmg

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-Interface-PR943-8ea3263-_32c3725b.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-8ea3263-_32c3725b.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-8ea3263-.tar_32c3725b.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-8ea3263-.tar_32c3725b.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR943/Vircadia-PR943-8ea3263-_32c3725b.exe

**Review by ctrlaltdavid — APPROVED** [review]

**Pull request merged**